Games/Retro_ML.SuperMarioBros/Game/SMBDataFetcher.cs

```diff
@@ -0,0 +1,453 @@
+        }
+
+        /// <summary>
+        /// Needs to be called every frame to reset the memory cache
+        /// </summary>
+        public void NextFrame()
+        {
+            frameCache.Clear();
+            internalClock.NextFrame();
+            InitFrameCache();
+        }
+
+        /// <summary>
+        /// Needs to be called every time a save state was loaded to reset the global cache.
+        /// </summary>
+        public void NextState()
+        {
+            frameCache.Clear();
+
+            tileCache.Clear();
+
+            levelCache.Clear();
+
+            internalClock.Reset();
+
+            currScreen = 0;
+        }
+
+        public byte[] GetTileArray() => Read(GameAddresses.Tiles);
+        public ushort GetCurrentScreen() => ReadSingle(GameAddresses.CurrentScreen);
+        public ushort GetMarioPositionX() => (ushort)(ReadSingle(PlayerAddresses.MarioPositionX) + (GetCurrentScreen() * 0x100));
+        public byte GetMarioPositionY() => ReadSingle(PlayerAddresses.MarioPositionY);
+        public ushort GetMarioScreenPositionX() => ReadSingle(PlayerAddresses.MarioScreenPositionX);
+        public ushort GetMarioScreenPositionY() => ReadSingle(PlayerAddresses.MarioScreenPositionY);
+        public byte[] IsSpritePresent() => Read(SpriteAddresses.IsSpritePresent);
+        public byte[] GetSpriteHitbox() => Read(SpriteAddresses.SpriteHitbox);
+        public byte[] GetSpritePositionX() => Read(SpriteAddresses.SpritePositionX);
+        public byte[] GetSpritePositionY() => Read(SpriteAddresses.SpritePositionY);
+        public byte[] GetSpriteScreenPosition() => Read(SpriteAddresses.SpriteScreenPosition);
+        public byte[] GetFirebarAngle() => Read(SpriteAddresses.FirebarSpinAngle);
+        public byte[] GetSprites() => Read(SpriteAddresses.SpriteType);
+        public byte[] IsHammerPresent() => Read(SpriteAddresses.IsHammerPresent);
+        public byte[] GetHammerHitbox() => Read(SpriteAddresses.HammerHitbox);
+        public ushort IsPowerUpPresent() => ReadSingle(SpriteAddresses.IsPowerUpPresent);
+        public byte[] GetPowerUpHitbox() => Read(SpriteAddresses.PowerUpHitbox);
+        public bool IsOnGround() => ReadSingle(PlayerAddresses.MarioFloatState) == 0x00;
+        public bool IsInWater() => ReadSingle(PlayerAddresses.IsSwimming) == 0x00;
+        public bool CanAct() => ReadSingle(PlayerAddresses.MarioActionState) == 0x08;
+        public bool IsDead() => ReadSingle(PlayerAddresses.MarioActionState) == 0x0B || ReadSingle(PlayerAddresses.IsFallingToDeath) == 0x01;
+        public bool WonLevel() => ReadSingle(PlayerAddresses.MarioActionState) == 0x04 || ReadSingle(PlayerAddresses.MarioActionState) == 0x05 || ReadSingle(GameAddresses.WonCondition) == 0x02 || ReadSingle(PlayerAddresses.MarioFloatState) == 0x03;
+        public bool IsAtMaxSpeed() => IsInWater() ? ReadSingle(PlayerAddresses.MarioMaxVelocity) == 0x18 : ReadSingle(PlayerAddresses.MarioMaxVelocity) == 0x28;
```

Comment:
Wouldn't this always be true, since we're reading Mario's Max velocity instead of reading Mario's current velocity?